### DISCUSSION TITLE
Packaging steps for distribution

### DISCUSSION THREAD

**Discussion by alexanderilyin** [behave/behave#1217] [Q&A] [answered]
Hi, love the tool, I wonder if there recommended way to distribute steps implementations (Python files)? I understand that there are a lot of project specific steps, but there are also some universal such as "Run sub-process and validate output" and I wonder what's the best way to extract those and share between repos?

**Comment by alexanderilyin** [comment]
I guess I just package Python files publish them to pypi.

**Comment by jenisys** [comment, marked as answer]
You basically have two choices as step writer:

* CASE 1: The steps are bundled with your feature files in the `steps` directory
* CASE 2: The steps are provided in a Python package and are used in a `step-library` with some glue code in the `steps` directory to import the needed/used steps from the Python package

If you want to reuse your steps in multiple test projects, you basically have CASE 2 (which is the focus point in this issue).
But this requires that your step library is generic enough to be used in multiple test contexts.

GLUE CODE:

* The glue code is currently required (as user control point).
* It allows the user to select which steps of a Python package are used and which are not used.
* It avoids the ambiguous-step problem that may occur.

**SEE ALSO:**
* [features/step.use_step_library.feature](https://github.com/behave/behave/blob/main/features/step.use_step_library.feature)
* EXAMPLE: [features/steps/use_steplib_behave4cmd.py](https://github.com/behave/behave/blob/main/features/steps/use_steplib_behave4cmd.py)

OTHER ASPECTS:

* You normally do not have a Python package that just contain steps.
* The steps often need some test infrastructure code (in other parts of this Python package) that makes writing the steps simpler and not overly complex.
* The test infrastructure parts help the to interact with the system that should be tested or perform tasks to simplify the steps.
* Therefore, the test infrastructure parts should be tested (as well as the steps) as normal for any Python package parts.

**Reply by alexanderilyin** [reply]
Looked around a little bit more and found `behave4cmd0` which is very close to what I was looking for.

**Reply by jenisys** [reply]
Note that `behave4cmd0` can be installed via `pip install git+https://github.com/behave/behave4cmd0@v1.3.1` (or `v1.4.0`) until it is published on pypi.
